### PR TITLE
TEPS 2022 volume extracts

### LINKED ISSUE #80: Extract volumes for 2022
- [ ] Miovision ATR data to the centreline
- [ ] ATR counts from MOVE to the centreline
- [ ] Vehicle Detector Stations (to the centreline)

We can create a new folder `volume_project/extract/2022` and put the `.sql` files in there.

### PR DESCRIPTION
**Pull request by scann0n**
Closes #80

We extracted short term ATR counts, Miovision ATR counts and RESCU counts for TEPS.

The readme explains the process (and hopefully provides some helpful guidance for next year too...)

Remember: since this is 2022, we used `gis.centreline_20220705`

Can you please look at the code for all of the sqls. Please pay particular attention to:
- the rescu detector id to centreline id matching. I did check every. single. location. but maybe I missed something. There were also 19 that I entered manually. The manually entered ones have a `null` value for `geom`
- do you agree with the rescu volume thresholds? If so, awesome, if not, let's talk
- extracting data from short term ATR counts - pretty straightforward but I don't have a ton of experience with this. Also I used `count_info_id` as the `volume_id` and I'm questioning my life choices.
- excluding questionable data from the Miovision extract using the anomalous ranges table (it's my first time!!!)

### CODE REVIEW

File: volume_project/extract/2022/rescu_sqls/01_lane_stats.sql [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            di.det_group,
+            v.entity_location_uid,
+            v.vdsconfig_uid,
+            di.direction,
+            vc.lanes,
+            UPPER(el.main_road_name) || ' and ' || UPPER(el.cross_road_name) AS gen_loc,
+            el.latitude,
+            el.longitude,
+            date_trunc('day', v.datetime_15min)::date AS dt,
+            COUNT(v.datetime_15min) AS bin_ct,
```
Comment (gabrielwol) [inline volume_project/extract/2022/rescu_sqls/01_lane_stats.sql:17]: In this case it was OK but in general you probably want to use `COUNT(DISTINCT ...)` for this - otherwise this just behaves like COUNT(*).

File: volume_project/extract/2022/rescu_sqls/01_lane_stats.sql [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    -- calculate daily volumes for days when detectors have data for every 15 minute bin 
+    WITH daily_vol AS (
+        SELECT
+            di.detector_id,
+            di.det_group,
+            v.entity_location_uid,
+            v.vdsconfig_uid,
+            di.direction,
+            vc.lanes,
+            UPPER(el.main_road_name) || ' and ' || UPPER(el.cross_road_name) AS gen_loc,
```
Comment (gabrielwol) [inline volume_project/extract/2022/rescu_sqls/01_lane_stats.sql:13]: I notice some results were missing cross_road_name like... `JARVIS RAMP TO W/B GARDINER and `. Maybe add a special case for this.

File: volume_project/extract/2022/readme.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+## Miovision Data
+There are two steps to this process:
+1. Rerun the miovision to centreline conflation table with the 2022-07-05 centreline data (see [this file](miovision_sqls/miovision_atr_2022.sql))
+2. Rerun this [data request](https://github.com/Toronto-Big-Data-Innovation-Team/bdit_data_requests/tree/master/volumes/atr/miovision/2023-02-03_UofT_miovision_to_centreline_update) with the 2022-07-05 miovision to centreline conflation table (see [this file](miovision_sqls/miovision_centreline_20220705.sql))
+
+After the data request is rerun, check to see if there are any instances where the `centreline_id` is null and the `volume` is greater than 0: 
+```
+SELECT * FROM mio_atr_centreline_20220705 WHERE centreline_id IS NULL AND volume > 0
+```
+There are also no notes in the `anomalous_range_notes` field, but miovision data that was marked as anomalous (where `project_level IN ('do-not-use', 'questionable')`) was omitted. The only non "do not use" or "questionable" entries in the anomalous range table in 2022 are for bicycles, which are not included in this request.
```
Comment (gabrielwol) [inline volume_project/extract/2022/readme.md]: Typo: ~~project_level~~ problem_level

File: volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql [hunk 4]
```diff
@@ -0,0 +1,173 @@
+-- this is a bit of a convoluded process... here are the steps I took. I'm happy to make it more efficient
```
Comment (gabrielwol) [inline volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql:1]: Do you want this to be a process I do when I add a new intersection? If so you can make a separate PR to update the [update intersections readme](https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/update_intersections/Readme.md)!

File: volume_project/extract/2022/readme.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+## Short Term ATR counts
+The table `traffic.arteries_centreline` is updated daily. Only 23 Of the unique arterycode and centreline_id combinations in 2022 did not have a `gis.centreline_20220705` equivalent (out of 1191), so no further centreline mapping was needed.
+
+Two types of short term counts were provided:
+- speed volume data was extracted using [this code](short_term_count_sqls/speedvol.sql)
+- classification data was extracted using [this code](short_term_count_sqls/classvol.sql)
+
+The table `scannon.oti_class` was created using [this code](short_term_count_sqls/oti_class.sql) to map classification id numbers to their descriptions.
```
Comment (gabrielwol) [inline volume_project/extract/2022/readme.md:45]: What was the source of this - can you link to it?

File: volume_project/extract/2022/short_term_count_sqls/oti_class.sql [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+(1, 'motorcycles'),
+(2, 'passenger cars'),
+(3, 'two axle, 4 tire single units'),
+(4, 'buses'),
+(5, 'two axis, 6 tire single units'),
+(6, 'three axle single units'),
+(7, 'four or more axle single units'),
+(8, 'four or less axle single trailers'),
+(9, 'five axle single trailers'),
+(10, 'six or more single trailers'),
```
Comment (gabrielwol) [inline volume_project/extract/2022/short_term_count_sqls/oti_class.sql]: 'six or more axle single trailers'

File: volume_project/extract/2022/short_term_count_sqls/speedvol.sql [hunk 7]
```diff
@@ -0,0 +1,22 @@
+CREATE MATERIALIZED VIEW scannon.speedvol_20220705 AS (
```
Comment (gabrielwol) [inline volume_project/extract/2022/short_term_count_sqls/speedvol.sql:1]: I noticed more than 2/3rds of this table (~5M rows) has volume = 0, is that necessary to include?

File: volume_project/extract/2022/short_term_count_sqls/classvol.sql [hunk 8]
```diff
@@ -0,0 +1,22 @@
+CREATE MATERIALIZED VIEW scannon.classvol_20220705 AS (
```
Comment (gabrielwol) [inline volume_project/extract/2022/short_term_count_sqls/classvol.sql:1]: Same comment as below about 2/3rds of the volumes being zero.

File: volume_project/extract/2022/readme.md (hunk 5 above)
Comment (gabrielwol) [reply volume_project/extract/2022/readme.md:45]: also should this reference table live outside of scannon?

File: volume_project/extract/2022/rescu_sqls/rescu_spdvol_centreline.sql [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    art.centreline_id, 
+    CASE
+        WHEN ev.direction = ANY (ARRAY['Eastbound'::text, 'Northbound'::text]) THEN 1
+        WHEN ev.direction = ANY (ARRAY['Westbound'::text, 'Southbound'::text]) THEN '-1'::integer
+        ELSE NULL::integer
+    END AS dir_bin,
+    v.datetime_15min AS count_bin,
+    v.count AS volume,
+    4 AS count_type,
+    v.speed_5kph AS speed_class,
```
Comment (gabrielwol) [inline volume_project/extract/2022/rescu_sqls/rescu_spdvol_centreline.sql:12]: Should we convert this speed_class to the same format as `prj_volume.speed_classes.speed_kph`? (or vice versa)

File: volume_project/extract/2022/rescu_sqls/rescu_counts_centreline.sql [hunk 10]
```diff
@@ -0,0 +1,20 @@
+CREATE TABLE scannon.rescu_count_centreline_20220705 AS (
```
Comment (gabrielwol) [inline volume_project/extract/2022/rescu_sqls/rescu_counts_centreline.sql:1]: Noticed there are some dupes here - multiple sensors mapped to the same centreline_id? We should select one with more reliable volumes - not sure if it will help but you could look at the recent data requests that [Adam](https://github.com/Toronto-Big-Data-Innovation-Team/bdit_data_requests/pull/571#issue-2064753848) or [I](https://github.com/Toronto-Big-Data-Innovation-Team/bdit_data_requests/pull/583/files) did which included sensor selection.
```sql
WITH dupes AS (
    SELECT centreline_id, dir_bin, count_bin, COUNT(*)
    FROM scannon.rescu_count_centreline_20220705
    GROUP BY centreline_id, dir_bin, count_bin
    HAVING COUNT(*) > 1
)

SELECT * FROM dupes
NATURAL JOIN scannon.rescu_count_centreline_20220705
ORDER BY centreline_id, dir_bin, count_bin
```

All the duplicate records are from these two centreline_ids:
1147358
30007492

Comment (gabrielwol) [reply volume_project/extract/2022/rescu_sqls/rescu_counts_centreline.sql:1]: This probably applies to the spdvol table as well. Could be worth adding primary keys on these tables to check all the joins aren't producing dupes!

File: volume_project/extract/2022/rescu_sqls/03_vol_threshold_check.sql [hunk 11]
```diff
@@ -0,0 +1,20 @@
+CREATE TABLE scannon.rescu_enuf_vol_22 AS (
```
Comment (gabrielwol) [inline volume_project/extract/2022/rescu_sqls/03_vol_threshold_check.sql:1]: I know how difficult sensor selection for rescu can be - don't mean to make you revisit everything, but I think there are some exceptions to these thresholds you should make in the top results here. Particularly the Gardiner east of Spadina (Jarvis / Sherbourne / Gardiner @ DVP / Yonge) - the volumes really drop off but seem valid to me based on traffic exiting into downtown. 

```
WITH vol_asmt AS (
    SELECT 
            *,
            CASE
                WHEN det_group = 'DVP' AND day_type = 'Weekday' AND d_vol / lanes  >= 15000 THEN 'enough volume'
                WHEN det_group = 'DVP' AND day_type = 'Weekend' AND d_vol / lanes  >= 10000 THEN 'enough volume'
                WHEN det_group = 'Gardiner' AND d_vol / lanes  >= 10000 THEN 'enough volume'
                WHEN det_group = 'Lakeshore' AND d_vol / lanes  >= 2000 THEN 'enough volume'
                
                ELSE 'check volume'
            END AS vol_check
        FROM scannon.rescu_dayvol_stats_22
        WHERE det_group <> 'On-Ramp'
    --WHERE detector_id IN ('DS0010DND', 'DS0010DSD')
    )
    
        SELECT detector_id, det_group, gen_loc, lanes, AVG(d_vol / lanes), COUNT(*)
        FROM vol_asmt WHERE vol_check = 'check volume'
        GROUP BY 1, 2, 3, 4
        ORDER BY COUNT(*) DESC
```

**Review by gabrielwol — COMMENTED** [review]
Some general comments! Not all of them need to be addressed for this request.
Overall: 
Miovision looks good!
ATR - omit zero records?
RESCU - thresholds look good to me, would just add some exceptions to eastern GArdiner as noted in one comment.

File: volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+            WHEN ld.nsew = 'NS' AND ld.latitude > ld.lat THEN 'N'
+            WHEN ld.nsew = 'NS' AND ld.latitude < ld.lat THEN 'S'
+            WHEN ld.nsew = 'EW' AND ld.longitude > ld.lng THEN 'E'
+            WHEN ld.nsew = 'EW' AND ld.longitude < ld.lng THEN 'W'
+        END AS leg
+    FROM ll_diff AS ld
+);
+
+--2. It's important that the centreline to miovision file has the right combination of intersection_uids and legs. Here's how I checked that:
+WITH data_legs AS (
```
Comment (chmnata) [inline volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql:96]: @gabrielwol Would the intersection_movement table be interchangeable for this as well?

Comment (gabrielwol) [reply volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql:96]: Yeah that would probably give the same result a lot quicker!

File: volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql [hunk 13]
```diff
@@ -0,0 +1,173 @@
+-- this is a bit of a convoluded process... here are the steps I took. I'm happy to make it more efficient
+-- 1. wrote code to find the centreline_ids that correspond to the segments that touch miovision intersections; put these in a table
+-- 2. did some error checking to see if that code worked (there were 9 situations where it didn't)...
+-- 3. manually investigated and fixed the centrelines that needed to be fixed (in the table)
+
+-- 1. Here is the code that finds centreline_ids that touch miovision intersections:
+CREATE TABLE scannon.centreline_miovision_20220705 AS (
```
Comment (chmnata) [inline volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql:7]:
```
select  intersection_uid, leg, count(1) 
from  scannon.centreline_miovision_20220705
group by intersection_uid, leg
having count(1) >1
```
returned four intersections + leg combination that got tagged to more than 1 centreline_id
92	"E"	2
71	"E"	2
91	"W"	2
97	"S"	2

File: volume_project/extract/2022/rescu_sqls/rescu_centreline.sql [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+CREATE TABLE scannon.rescu_cent_20220705 AS (
+
+-- get centreline segments
+WITH cl_segs AS (
+SELECT 
+    cl.geo_id AS centreline_id,
+    cl.lf_name AS cent_name,
+    ST_SetSRID(cl.geom, 4326) AS cl_geom
+FROM gis.centreline_20220705 AS cl
+WHERE fcode_desc IN('Major Arterial', 'Expressway') -- these are the only types of roads we need
```
Comment (chmnata) [inline volume_project/extract/2022/rescu_sqls/rescu_centreline.sql:11]: no ramps and stuff?

File: volume_project/extract/2022/rescu_sqls/rescu_centreline.sql [hunk 15]
```diff
@@ -0,0 +1,161 @@
+-- Make a table that links RESCU detectors with the centreline from 2022-07-05
+CREATE TABLE scannon.rescu_cent_20220705 AS (
```
Comment (chmnata) [inline volume_project/extract/2022/rescu_sqls/rescu_centreline.sql:2]:
```
select  * from scannon.rescu_cent_20220705  
where cl_geom is null
```
There are 19 rows with no geom but with a centreline_id!

File: volume_project/extract/2022/rescu_sqls/rescu_centreline.sql [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+
```
Comment (chmnata) [inline volume_project/extract/2022/rescu_sqls/rescu_centreline.sql:3]: image
`DW0020DWL` can you double check if this is a correct match ?

File: volume_project/extract/2022/rescu_sqls/rescu_centreline.sql (hunk 15 above)
Comment (chmnata) [reply volume_project/extract/2022/rescu_sqls/rescu_centreline.sql:2]: I think those are the ones you manually inserted, can you do a update set geom for those ones?

File: volume_project/extract/2022/rescu_sqls/rescu_centreline.sql [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+-- get centreline segments
```
Comment (chmnata) [inline volume_project/extract/2022/rescu_sqls/rescu_centreline.sql:4]: Question! 
For rescu locations like this, do we want to join to both centreline or just one ?
image

Comment (radumas) [reply volume_project/extract/2022/rescu_sqls/rescu_centreline.sql:4]: I don't know if the model would overweight counts for this sensor if its data was assigned to 2 centrelines

File: volume_project/extract/2022/rescu_sqls/rescu_centreline.sql (hunk 16 above)
Comment (gabrielwol) [reply volume_project/extract/2022/rescu_sqls/rescu_centreline.sql:3]: I think I recognize this as a case where there were two sensors with the same description and I picked the one with the higher volume when looking at it last time.

File: volume_project/extract/2022/readme.md (hunk 5 above)
Comment (scann0n) [reply volume_project/extract/2022/readme.md:45]: Here is the [source on Notion](https://www.notion.so/bditto/Feature-Classification-ATRs-27ece0049d654c9ba06136bffc07e2e8?pvs=4#1fbcca593276484a9ecfbf550f7cc4aa). I asked for assistance on the data talk slack channel and Maddy sent me me a link.

I agree that it shouldn't be on `scannon`. Do we trust it enough to put it somewhere like `prj_volume` or `traffic`? I did very mild QC (yup, cars are the biggest category, seems legit).

File: volume_project/extract/2022/short_term_count_sqls/oti_class.sql (hunk 6 above)
Comment (scann0n) [reply volume_project/extract/2022/short_term_count_sqls/oti_class.sql]: Implemented!

File: volume_project/extract/2022/short_term_count_sqls/speedvol.sql (hunk 7 above)
Comment (scann0n) [reply volume_project/extract/2022/short_term_count_sqls/speedvol.sql:1]: I added `WHERE sv.count > 0`

File: volume_project/extract/2022/short_term_count_sqls/classvol.sql (hunk 8 above)
Comment (scann0n) [reply volume_project/extract/2022/short_term_count_sqls/classvol.sql:1]: I added `WHERE sv.count > 0`

File: volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql (hunk 4 above)
Comment (scann0n) [reply volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql:1]: GREAT IDEA!!! Will do!

File: volume_project/extract/2022/rescu_sqls/rescu_counts_centreline.sql (hunk 10 above)
Comment (scann0n) [reply volume_project/extract/2022/rescu_sqls/rescu_counts_centreline.sql:1]: I ran `SELECT * FROM scannon.rescu_cent_20220705 WHERE centreline_id IN (1147358, 30007492)` and got 2 detectors on 1147358 and 2 on 30007492 (4 detectors in total).

So for 1147358, both detectors are on one centreline segment:
image

The segment is quite long:
image

It's the same deal with 30007492 - both detectors even have the same "east of Parkside" descriptor even!
image

And the segment:
image

File: volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql (hunk 13 above)
Comment (scann0n) [reply volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql:7]: I checked these out. 

For 3 of them (92, 71 and 97) I updated the direction to match the (correct) `centreline_id` (in other words, where two legs had been given an 'E', one should've been an 'S', so I updated the table accordingly.

For 91 (Lakeshore and Spadina), there are actually 5 centreline segments and two are west legs - one is Lakeshore and the other is the ramp to the Gardiner. How is this handled in other situations? This is a relatively new intersection.

File: volume_project/extract/2022/rescu_sqls/rescu_centreline.sql (hunk 14 above)
Comment (scann0n) [reply volume_project/extract/2022/rescu_sqls/rescu_centreline.sql:11]: No ramps allowed!

File: volume_project/extract/2022/rescu_sqls/rescu_centreline.sql (hunk 16 above)
Comment (scann0n) [reply volume_project/extract/2022/rescu_sqls/rescu_centreline.sql:3]: `DW0020DWL` is a Lakeshore detector (near Lakeshore and Bay), and it's mapped to the Lakeshore segment that's just west of Bay.

File: volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql (hunk 13 above)
Comment (scann0n) [reply volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql:7]: ...and since it's new it wasn't around in 2022 so I ran the miovision mat view with the updated `scannon.miovision_centreline_20220705` table.

File: volume_project/extract/2022/rescu_sqls/rescu_counts_centreline.sql (hunk 10 above)
Comment (gabrielwol) [reply volume_project/extract/2022/rescu_sqls/rescu_counts_centreline.sql:1]: I suggest eliminating the ones with lower volumes

Comment (scann0n) [reply volume_project/extract/2022/rescu_sqls/rescu_counts_centreline.sql:1]: I determined how many days each detector had of good data, and the results are:
image
So `DW0070DEG` and `DW0080DEG` are mapped to the same `centreline_id` and `DW0080DEG` has more good days of data so we'll use  `DW0080DEG`.

Of the other pair, `DW0126DEG` has way more days of data so we'll go with that one.

File: volume_project/extract/2022/rescu_sqls/rescu_spdvol_centreline.sql (hunk 9 above)
Comment (scann0n) [reply volume_project/extract/2022/rescu_sqls/rescu_spdvol_centreline.sql:12]: IMO we shouldn't lose the finer grain data source. Classifications are also inconsistent between data sources.

File: volume_project/extract/2022/rescu_sqls/rescu_counts_centreline.sql (hunk 10 above)
Comment (scann0n) [reply volume_project/extract/2022/rescu_sqls/rescu_counts_centreline.sql:1]: Reran the `WITH dupes AS` code block above and returned zero results so yay!

File: volume_project/extract/2022/rescu_sqls/03_vol_threshold_check.sql (hunk 11 above)
Comment (scann0n) [reply volume_project/extract/2022/rescu_sqls/03_vol_threshold_check.sql:1]: Considered changing the threshold to 6000 per lane per day... discussed with @radumas and he said we should keep the thresholds where they are since they are already low.

File: volume_project/extract/2022/rescu_sqls/01_lane_stats.sql (hunk 1 above)
Comment (scann0n) [reply volume_project/extract/2022/rescu_sqls/01_lane_stats.sql:17]: GOOD TO KNOW!!!

File: volume_project/extract/2022/readme.md (hunk 3 above)
Comment (scann0n) [reply volume_project/extract/2022/readme.md]: Updated!

File: volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql (hunk 4 above)
Comment (scann0n) [reply volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql:1]: General Improvement 2: update the miovision to centreline table when we add new intersections!

File: volume_project/extract/2022/readme.md (hunk 5 above)
Comment (scann0n) [reply volume_project/extract/2022/readme.md:45]: General Improvement Idea 3: add the classification table to `traffic` (and maybe do some more QC)

File: volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql (hunk 12 above)
Comment (scann0n) [reply volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql:96]: Incorporate this thought into General Improvement 2 (updating the miovision to centreline table when we add new intersections).

File: volume_project/extract/2022/rescu_sqls/rescu_centreline.sql (hunk 15 above)
Comment (scann0n) [reply volume_project/extract/2022/rescu_sqls/rescu_centreline.sql:2]: Yes, the 19 missing geoms are the ones I updated manually (I wanted to be able to identify them quickly). This is an intermediary table that only includes RESCU detectors with good data in 2022, so I think this should be implemented via

General Improvement Idea 4: make a RESCU to centreline equivalency table (consider expanding this to include other VDS types)

File: volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql (hunk 13 above)
Comment (chmnata) [reply volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql:7]: if its lakeshore and spadina then i guess lakeshore ? 🤔

File: volume_project/extract/2022/rescu_sqls/rescu_centreline.sql (hunk 15 above)
Comment (chmnata) [inline volume_project/extract/2022/rescu_sqls/rescu_centreline.sql:2]: There are 14 matches where the direction from detector doesn't match with the centreline name. I think it could be sometimes one 1 centreline represents both direction..... do you mind checking them to make sure thats the case?

```
select direction, detector_id, gen_loc, lf_name, right(lf_name, 1), left(direction, 1)
from scannon.rescu_cent_20220705 a
inner join gis.centreline_20220705 on centreline_id = geo_id
where right(lf_name, 1)!= left(direction, 1)
```

Also there is one gen_loc named `and` 😆 , detector_id `DW0151DEG`

File: volume_project/extract/2022/rescu_sqls/01_lane_stats.sql (hunk 2 above)
Comment (gabrielwol) [reply volume_project/extract/2022/rescu_sqls/01_lane_stats.sql:13]: My one remaining comment - I think it's the same as Natalie's: https://github.com/CityofToronto/bdit_volumes/pull/81#discussion_r1554076676

**Review by gabrielwol — APPROVED** [review]
Thanks for making all those changes!

File: volume_project/extract/2022/rescu_sqls/rescu_centreline.sql (hunk 15 above)
Comment (scann0n) [reply volume_project/extract/2022/rescu_sqls/rescu_centreline.sql:2]: Yes I will certainly double check!!!

And also, the east bound lanes of Lakeshore Blvd West and the west bound lanes of Lakeshore Blvd East caused me some :meow_headache:

Yes that `and` should be taken care of via General Improvement Idea 1...

File: volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql (hunk 4 above)
Comment (gabrielwol) [reply volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql:1]: Issue created: https://github.com/CityofToronto/bdit_data-sources/issues/922

File: volume_project/extract/2022/rescu_sqls/rescu_centreline.sql (hunk 15 above)
Comment (gabrielwol) [reply volume_project/extract/2022/rescu_sqls/rescu_centreline.sql:2]: New issue: https://github.com/CityofToronto/bdit_data-sources/issues/923

Comment (scann0n) [reply volume_project/extract/2022/rescu_sqls/rescu_centreline.sql:2]: K I have checked them all and the locations are correct! And they were all on Lake Shore...

Here's how I checked:
- Looked at the rescu geom
- Looked at the corresponding segment
- Looked at the centreline for `lf_name LIKE '%Lake Shore%'` so that I could see if Lake Shore was represented with 1 or 2 lines
    - If Lake Shore was represented with 2 lines, I made sure Eastbound detectors were mapped to the southern segment and that Westbound detectors were mapped to the northern segment

File: volume_project/extract/2022/short_term_count_sqls/tmc_2_atr_vol.sql [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+        WHEN from_dir = ANY (ARRAY['WB'::text, 'SB'::text]) THEN '-1'::integer
+        ELSE NULL::integer
+    END AS dir_bin,
+    datetime_bin AS count_bin,
+    volume,
+    3 AS count_type,
+    NULL AS speed_class,
+    traffic_classification AS vehicle_class,
+    count_info_id AS volume_id
+FROM vol_cent
```
Comment (gabrielwol) [inline volume_project/extract/2022/short_term_count_sqls/tmc_2_atr_vol.sql:41]: I noticed in bigdata you've since updated this to include `WHERE vol_cent.tcl_from_segment IS NOT NULL` (and similar below). Can you please update the query here?
Noting that this excludes about 7% of the volume.

File: volume_project/extract/2022/rescu_sqls/01_lane_stats.sql (hunk 2 above)
Comment (scann0n) [reply volume_project/extract/2022/rescu_sqls/01_lane_stats.sql:13]: There is one entry that's just `and`!

General improvement idea 1: do some QC on the `vds.entity_locations.main_road_name` and `cross_road_name` fields!

File: volume_project/extract/2022/short_term_count_sqls/tmc_2_atr_vol.sql (hunk 18 above)
Comment (scann0n) [reply volume_project/extract/2022/short_term_count_sqls/tmc_2_atr_vol.sql:41]: Updated!

That 7% is going into non-centreline locations like mall entrances

File: volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql (hunk 13 above)
Comment (gabrielwol) [reply volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql:7]: This still won't affect TEPS 2023 because the intersection was only up and running in 2024, but yeah... weird intersection. Both of those western legs are measured by the camera.
image

Comment (scann0n) [reply volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql:7]: Wut...
We need a way to distinguish between those two legs (north and south western legs?)
